proposals/004_Overlays.md

```diff
@@ -0,0 +1,178 @@
+
+Field Name | Type | Description
+---|:---:|---
+<a name="updateTarget"></a>target | `string` | A JMESPath expression referencing the target objects in the target document.
+<a name="updateValue"></a>value | [Any](#valueObject) | An object with the properties and values to be updated in the target document.  Property has no impact if `remove` property is `true`.
+<a name="updateRemove"></a>remove | `boolean` | A boolean value that indicates that the target object is to be removed from the the map or array it is contained in. The default value is false.  
+
+The properties of the `Value Object` MUST be compatible with the target object referenced by the JMESPath key.  When the Overlay document is applied, the properties in the `Value Object` replace properties in the target object with the same name and new properties are appended to the target object.
+
+
+##### Structured Overlays Example
+
+When updating properties throughout the target document it may be more efficient to create a single `Update Object` that mirrors the structure of the target document. e.g.
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Structured Overlay
+  version: 1.0.0
+updates:
+- target: "@"
+    value:
+      info:
+        x-overlay-applied: structured-overlay
+      paths:
+        "/":
+          summary: "The root resource"
+          get:
+            summary: "Retrieve the root resource"
+            x-rate-limit: 100
+        "/pets":
+          get:
+            summary: "Retrieve a list of pets"
+            x-rate-limit: 100
+      components:
+      tags:
+```
+
+##### Targeted Overlays
+
+Alternatively, where only a small number of updates need to be applied to a large document, each [Update Object](#updateObject) can be more targeted.
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Structured Overlay
+  version: 1.0.0
+updates:
+- target: paths."/foo".get
+    value:
+        description: This is the new description
+- target: paths."/bar".get
+    value:
+        description: This is the updated description
+- target: paths."/bar"
+    value:
+        post:
+            description: This is an updated description of a child object
+            x-safe: false
+```
+
+##### Wildcard Overlays Examples
+
+One significant advantage of using the JMESPath syntax that it allows referencing multiple nodes in the target document.  This would allow a single update object to be applied to multiple target objects using wildcards.
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Update many objects at once
+  version: 1.0.0
+updates:
+- target: paths.*.get
+    value:
+      x-safe: true
+- target: paths.*.get.parameters[?name=='filter' && in=='query']
+    value:
+      schema:
+        $ref: "/components/schemas/filterSchema"
+```
+
+##### Array Modification Examples
+
+Due to the fact that we can now reference specific elements of the parameter array, it does open the possibilty to being able to add parameters and potentially remove them using a `null` value.
```

Comment:
Is "using a null value" a hangover from a previous version without the `remove` property?

Reply:
We should probably also document what happens to array indices when an item is removed.

Consider the array: `[0, 1, 2, 3, 4]`

If we `remove` item `1`, and then remove item `2`, are we left with `[0, 3, 4]` or `[0, 2, 4]`?

Reply:
Yes. The `null` language is left over.  I will remove it.

Yes, it would be good to clarify the behavior around remove.  My understanding is the second case is true.  All Updates are applied sequentially and use the result of the previous update as the baseline.  If we don't do that then I think we could get into some funky merge situations even within a single document.

Reply:
Thanks for the feedback